packages/contracts/src/contracts/current/protocol/Exchange/MixinWrapperFunctions.sol

```diff
@@ -16,281 +16,259 @@
-            mstore(add(start, 260), add(orderValues, 96))   // takerFee
-            mstore(add(start, 292), add(orderValues, 128))  // expiration
+            // Load free memory pointer
+            let start := mload(0x40)
+
+            // Write function signature
+            mstore(start, fillOrderSelector)
+
+            // Write order struct
+            mstore(add(start, 4), mload(order))             // makerAddress
+            mstore(add(start, 36), mload(add(order, 32)))   // takerAddress
+            mstore(add(start, 68), mload(add(order, 64)))   // makerTokenAddress
+            mstore(add(start, 100), mload(add(order, 96)))  // takerTokenAddress
+            mstore(add(start, 132), mload(add(order, 128))) // feeRecipientAddress
+            mstore(add(start, 164), mload(add(order, 160))) // makerTokenAmount
+            mstore(add(start, 196), mload(add(order, 192))) // takerTokenAmount
+            mstore(add(start, 228), mload(add(order, 224))) // makerFeeAmount
+            mstore(add(start, 260), mload(add(order, 256))) // takerFeeAmount
+            mstore(add(start, 292), mload(add(order, 288))) // expirationTimeSeconds
+            mstore(add(start, 324), mload(add(order, 320))) // salt
+
+            // Write takerTokenFillAmount
             mstore(add(start, 356), takerTokenFillAmount)
+
+            // Write signature offset
             mstore(add(start, 388), 416)
-            mstore(add(start, 420), signatureLength)
-        }
-        
-        // Write signature contents and padding one byte at a time
-        for (uint256 i = 0; i < signatureLength; i++) {
-            input[452 + i] = signature[i];
-        }
-        for (i = 0; i < paddingLength; i++) {
-            input[452 + signatureLength + i] = 0x00;
-        }
-        
-        // Call the function
-        assembly {
-            success := delegatecall(
-                gas,       // If fillOrder `revert()`s, we recover unused gas
-                address,   // call this contract
-                start,     // pointer to start of input
-                inputSize, // length of input
-                start,     // store output over input
-                32         // output is 32 bytes
+
+            // Write signature length
+            let sigLen := mload(signature)
+            mstore(add(start, 420), sigLen)
+
+            // Calculate signature length with padding
+            let paddingLen := sub(32, mod(sigLen, 32))
+            let sigLenWithPadding := add(sigLen, paddingLen)
+
+            // Write signature
+            let sigStart := add(signature, 32)
+            for { let curr := 0 } 
+            lt(curr, sigLenWithPadding)
+            { curr := add(curr, 32) }
+            { mstore(add(start, add(452, curr)), mload(add(sigStart, curr))) }
+
+            // Execute delegatecall
+            let success := delegatecall(
+                gas,                         // forward all gas 
```

Comment:
In some cases a call can fail consuming all gas. This should be rare now that Solidity reverts by default.

In [0.4.9 (17-01-03)](https://github.com/ethereum/solidity/blob/develop/Changelog.md#049-2017-01-31) revert became default for internal errors, but still all gas consumed for `throw`.

The changelog is not clear on wheter this is still the current behaviour. It is also unclear on what `assert` and `require` do. (It says they 'throw' but also that they are an 'internal error').

We should research this and document the decision to forward all gas and the consequences.

Reply:
I'll add this as a TODO